src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs

```diff
@@ -145,46 +153,48 @@ internal string CreateCacheKey(CachedVaryBy varyBy)
             var builder = new StringBuilder()
                 .Append(request.Method.ToUpperInvariant())
                 .Append(";")
-                .Append(request.Path.Value.ToUpperInvariant())
-                .Append(CreateVaryByCacheKey(varyBy));
+                .Append(request.Path.Value.ToUpperInvariant());
 
-            return builder.ToString();
-        }
-
-        private string CreateVaryByCacheKey(CachedVaryBy varyBy)
-        {
-            // TODO: resolve key format and delimiters
-            if (varyBy == null || varyBy.Headers.Count == 0)
+            if (varyBy?.Headers.Count > 0)
             {
-                return string.Empty;
-            }
-
-            var builder = new StringBuilder(";");
+                // TODO: resolve key format and delimiters
+                foreach (var header in varyBy.Headers)
+                {
+                    // TODO: Normalization of order, case?
+                    var value = HttpContext.Request.Headers[header];
 
-            foreach (var header in varyBy.Headers)
-            {
-                // TODO: Normalization of order, case?
-                var value = HttpContext.Request.Headers[header].ToString();
+                    // TODO: How to handle null/empty string?
+                    if (StringValues.IsNullOrEmpty(value))
+                    {
+                        value = "null";
```

Comment:
If the user explicitly specified to vary by a header, we use a placeholder value of `null` to indicate a missing value. This differentiates between an empty value of a request with an explicit varyby verses a request that does not varyby that header. May revisit this decision which is why I added a TODO item here. Maybe it makes sense to make this behaviour configurable.